hovercraft/__init__.py

```diff
@@ -121,57 +126,67 @@ def my_gettext(s):
     if args.template and args.template not in ('simple', 'default'):
         args.template = os.path.abspath(args.template)
 
-    if args.targetdir:
+    args.basedir = os.getcwd()
+
+
+    if args.targetdir and not os.path.exists(args.targetdir):
+        os.makedirs(args.targetdir)
+
+    if args.build:
         # Generate the presentation
         generate(args)
     else:
         # Server mode. Start a server that serves a temporary directory.
+        if args.targetdir:
+            start_server(args)
+        else:
+            with TemporaryDirectory() as targetdir:
+                args.targetdir = targetdir
+                start_server(args)
+
+def start_server(args):
+    args.presentation = os.path.abspath(args.presentation)
+
+    # Set up watchdog to regenerate presentation if saved.
+    event = threading.Event()
+    event.set()
+    thread = threading.Thread(target=generate_and_observe, args=(args, event))
+    try:
+        # Serve presentation
+        if ':' in args.port:
+            bind, port = args.port.split(':')
+        else:
+            bind, port = '0.0.0.0', args.port
+        port = int(port)
+
+        # First create the server. This checks that we can connect to
+        # the port we want to.
+        server = HTTPServer((bind, port), SimpleHTTPRequestHandler)
+        print("Serving HTTP on", bind, "port", port, "...")
+
+        try:
+            # Now generate the presentation
+            thread.start()
 
-        with TemporaryDirectory() as targetdir:
-            args.targetdir = targetdir
-            args.presentation = os.path.abspath(args.presentation)
-
-            # Set up watchdog to regenerate presentation if saved.
-            event = threading.Event()
-            event.set()
-            thread = threading.Thread(target=generate_and_observe, args=(args, event))
             try:
-                # Serve presentation
-                if ':' in args.port:
-                    bind, port = args.port.split(':')
-                else:
-                    bind, port = '0.0.0.0', args.port
-                port = int(port)
-
-                # First create the server. This checks that we can connect to
-                # the port we want to.
-                os.chdir(targetdir)
```

Comment:
I'm guessing that the problem is because of this chdir, it must change the current directory for the thread as well, which I didn't expect, although I haven't verified that. But since you remove this, then the server mode can not longer reasonably work.